core/controllers/base.py

```diff
@@ -164,7 +164,12 @@ def __init__(self, request, response):  # pylint: disable=super-init-not-called
             self.payload = None
         self.iframed = False
 
-        auth_claims = auth_services.get_auth_claims_from_request(request)
+        try:
+            auth_claims = auth_services.get_auth_claims_from_request(request)
+        except ValueError:
```

Comment:
This seems slightly hacky--can we instead destroy the auth session at the time of revoking privileges rather than on an error? How can we be sure this error actually warrants destroying the auth session vs. just triggering an HTTP failure--couldn't a bunch of implementation scenarios trigger a ValueError?

Reply:
This actually catches different kinds of errors, and not catching an exception here would cause the entire server to fail with a 500 error.

Furthermore, auth sessions refer to the currently logged-in user so, in the case of revoking privileges, this is the super admin. However, we want to destroy the session of the _targeted user_, not the super admin, so that doesn't make sense to do here.

This try/catch intends to catch issues I've faced when managing cookies on the development server. When I shut down the server without logging out and destroying the session, the next time I start the server it'd be flooded by "FirebaseError: User does not exist" errors (which makes sense; the server has been wiped out). This code acknowledges that cookies might have invalid values, and destroys such sessions as soon as they're detected rather than allowing them to persist.

Please let me know if that makes sense.

Reply:
I've added a note explaining the intent to the code, PTAL!

Reply:
Thanks for the comment. I'm still unsure about this, I think.

First, this issue doesn't actually crash any server instances, does it? I thought webapp catches the error so that it can issue an http error code (such as 500), and gracefully recovers. Is that not the case?

My concern here is that catching ValueError seems a bit like a sledgehammer when trying to specifically work around a very specific issue: invalid auth cookie. Is there not a better way in which we can check this? It's generally frowned down to use exception handling for control flow, so I'm really hoping there's a better way to directly check the validity of the cookie rather than potentially catching and/or silencing legitimate errors that also trigger ValueError (& should result in a 500).

Reply:
So the actual behavior is a 500 error that causes the page to fail completely. You can actually see this behavior in develop by signing in and then restarting the server: the page will fail until we manually clear the cookies.

Regarding the concern, an exception is raised by the Firebase SDK specifically when the cookie is invalid, so I transform it into a "simpler" error in the platform layer for `base.py` to catch. Also, I'm not really convinced that an exception doesn't fit here, an invalid session cookie seems as exception-worthy as any other function that throws a ValueError; an invalid value was detected.

Reply:
I've updated the comment to reflect these thoughts, PTAL!

Reply:
I think there might be a slight miscommunication here. To be clear: I don't see catching the ValueError as the same as detecting invalid session cookie. I see it as catching a bunch of different potential failures, one of which is invalid session cookie. I agree that detecting an invalid session cookie is an exceptional case and should be handled as such, but I disagree with catching a generic exception as the manner to do it since we can accidentally catch unrelated errors that would be valid to result in a 500 error.

I'm guessing the ValueError is coming from this: https://github.com/oppia/oppia/blob/7964a49d527164a567c51021b9757bca60b7e683/core/platform/auth/firebase_auth_services.py#L531. Is that correct? If so, could we just check in that function whether the cookie is present and, if not, return an error rather than throw? This will result in more robustness since it explicitly communicates to the caller that an invalid session error occurred & that it should be specifically handled.

Let me know if you want to chat over Hangouts if that would be more expedient.

Reply:
The error comes from the following line: https://github.com/oppia/oppia/blob/d08dec214057916a204655baf26d20a42b2270e2/core/platform/auth/firebase_auth_services.py#L624

Returning an error seems worse than raising an exception, I'm still convinced that using Python exceptions is the correct and idiomatic approach here. Instead, perhaps creating a custom exception type would be better? Something like `InvalidSessionError`?

Reply:
> I'm still convinced that using exceptions is the correct and idiomatic approach here, returning an error object feels like an anti-pattern we should avoid

Conventional programmer wisdom actually suggests the opposite, I think. See:
- https://stackoverflow.com/a/729404
- https://stackoverflow.com/a/729412
- https://wiki.c2.com/?DontUseExceptionsForFlowControl
- https://ahdak.github.io/blog/effective-java-part-9/ (Java specific)
- https://softwareengineering.stackexchange.com/a/351121 (particularly noteworthy since exceptions are more conventional in Python world)

(& many, many more references)

In general, exceptions should be reserved for exceptional cases. The typical pattern to solve these situations _is_ to use a return value, not an exception (even in the Python world where exceptions are more permitted).

I don't want to push too hard against the general idiom of using exceptions for control flow despite my personal thoughts on it, but using a general exception is particularly harmful & using a custom or more specific exception is one mitigation for that. I'm definitely happier with the current approach, but I think it might be prudent for @seanlip to weigh in on the potential design pattern that we might unintentionally and explicitly be setting precedent on here: whether to use return values as is conventional in most systems & languages, or exceptions which seems more conventional in Oppia (I'm assuming we want the latter for consistency, but we should be explicit about understanding the trade-offs here).

I'll review the PR itself as a follow-up to this comment.

Reply:
Hmm, I'm confused why you think this counts as flow control? A cookie becomes invalid/expires in rare situations, hence why an exception is raised. The links you've posted refer to situations where the "exceptional" branch is very common (like `list_contains` raising an error to signal a missing value).

Reply:
@brianrodri -- just to confirm, are the two exceptional cases here ones which shouldn't be happening in practice?

Reply:
`StaleAuthSessionError` will happen in practice (every 2 weeks when the cookie expires). `InvalidAuthSessionError` is a bug that needs to be debugged.

Reply:
I thought about it and I think that, in this case, exceptions are fine. The added specificity of the exceptions definitely helps (I wouldn't have been happy with ValueError) and as a developer I understand what's going on in the code. So I don't think a change is needed.

I take the point about setting precedent, but I'm not particularly worried here. Many other functions raise exceptions for circumstances which could occur but are not "supposed" to. I think in this case it's OK since it's an uncommon flow and fits the intuitive definition of an exception case to me.

@BenHenning does this give you the info you need from my end?

Reply:
Yes it does, thanks @seanlip.